### PR TITLE
Fuse the input -> p2p buffer copy into one-shot all-reduce kernel when the input is small

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121213

This improves the gpt-fast llama2 70B 8xH100 (non-standard) TP benchmark from 86 tok/s to 88 tok/s.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/csrc/distributed/c10d/intra_node_comm.cu [hunk 1]
```diff
@@ -139,12 +139,21 @@ static __global__ void oneShotAllReduceKernel(
...
   const size_t offset =
       (blockDim.x * blockIdx.x + threadIdx.x) * numelPerThread;
   const size_t stride = blockDim.x * gridDim.x * numelPerThread;
 
+  if (fuseInputCopy) {
+    for (size_t i = offset; i < N_aligned; i += stride) {
+      bf16x8 val;
+      streamLoad128(val, &input[i]);
+      streamStore128(&buffers[rank][i], val);
+    }
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/intra_node_comm.cu:154]: IIUC, we are loading 8 x bf16 at one time.
But here `stride` seems to be at the unit of bf16. 
Am I missing anything?

Comment (yifuwang) [reply torch/csrc/distributed/c10d/intra_node_comm.cu:154]: Hi @kwen2501, the unit of `stride` is `numelPerThread = kBytesPerThread / sizeof(at::BFloat16)` which is also 8 x bf16.

**Review by Chillee — APPROVED** [review]

**Pull request closed without merging**